cmd/segment-reaper/observer_test.go

```diff
@@ -0,0 +1,494 @@
+			err := obsvr.processSegment(ctx.Context, objSeg.path, objSeg.pointer)
+			require.NoError(t, err)
+		}
+
+		// Assert observer internal state
+		assert.Equal(t, projectID.String(), obsvr.lastProjectID, "lastProjectID")
+		assert.Equal(t, 1, len(obsvr.objects), "objects number")
+		if assert.Contains(t, obsvr.objects, bucketName, "bucket in objects map") {
+			if assert.Equal(t, 1, len(obsvr.objects[bucketName]), "objects in object map") {
+				if assert.Contains(t, obsvr.objects[bucketName], objPath, "path in bucket objects map") {
+					obj := obsvr.objects[bucketName][objPath]
+					assert.Equal(t, numSegments, int(obj.expectedNumberOfSegments), "Object.expectedNumSegments")
+					assert.True(t, obj.hasLastSegment, "Object.hasLastSegment")
+					assert.False(t, obj.skip, "Object.skip")
+				}
+			}
+		}
+
+		// Assert observer global stats
+		assert.Zero(t, obsvr.inlineSegments, "inlineSegments")
+		assert.Zero(t, obsvr.lastInlineSegments, "lastInlineSegments")
+		assert.Equal(t, numSegments, obsvr.remoteSegments, "remoteSegments")
+	})
+
+	t.Run("objects with at least one has more than 65 segments", func(t *testing.T) {
+		ctx := testcontext.New(t)
+		defer ctx.Cleanup()
+
+		var (
+			testdata = generateTestdataObjects(ctx.Context, t, false, true)
+			obsvr    = observer{
+				objects: make(bucketsObjects),
+			}
+		)
+
+		for _, objSeg := range testdata.objSegments {
+			err := obsvr.processSegment(ctx.Context, objSeg.path, objSeg.pointer)
+			require.NoError(t, err)
+		}
+
+		// Assert observer internal state
+		assert.Equal(t, testdata.projectID.String(), obsvr.lastProjectID, "lastProjectID")
+		if assert.Equal(t, len(testdata.expectedObjects), len(obsvr.objects), "objects number") {
+			for bucket, bucketObjs := range obsvr.objects {
+				expBucketObjs, ok := testdata.expectedObjects[bucket]
+				if !ok {
+					t.Errorf("bucket '%s' shouldn't exist in objects map", bucket)
+					continue
+				}
+
+				if !assert.Equalf(t, len(expBucketObjs), len(bucketObjs), "objects per bucket (%s) number", bucket) {
+					continue
+				}
+
+				for expPath, expObj := range expBucketObjs {
+					if !assert.Contains(t, bucketObjs, expPath, "path in bucket objects map") {
+						continue
+					}
+
+					obj := bucketObjs[expPath]
+					assert.Equal(t, expObj.expectedNumberOfSegments, obj.expectedNumberOfSegments, "Object.expectedNumSegments")
+					assert.Equal(t, expObj.hasLastSegment, obj.hasLastSegment, "Object.hasLastSegment")
+
+					assert.Equal(t, expObj.skip, obj.skip, "Object.skip")
+					if !expObj.skip {
+						assert.Equal(t, expObj.segments, obj.segments, "Object.segments")
+					}
+				}
+			}
+		}
+
+		// Assert observer global stats
+		assert.Equal(t, testdata.expectedInlineSegments, obsvr.inlineSegments, "inlineSegments")
+		assert.Equal(t, testdata.expectedInlineSegments, obsvr.lastInlineSegments, "lastInlineSegments")
+		assert.Equal(t, testdata.expectedRemoteSegments, obsvr.remoteSegments, "remoteSegments")
```

Comment:
How are we sure that we are checking for the correct expected values here? What if we have the same bug in the `processSegment` implementation and in the `generateTestdataObjects` helper function?

Ideally, the expected values should be constants - something that can be easily validated with the naked eye.

I am worried that the code logic for setting up these test cases is significantly more complex than the implementation code it is testing. My understanding is that tests are simpler code that assert the correctness of more complex code.

Reply:
I commented that before. We can have all the cases in static values, but to represent all the combinations we need quite a bit effort, much less now after this exists.

When I had to take a look at the implementation and write a test it was hard for me to see that the implementation could work in any specific case, mostly because of the internal state that observer has and reset when the first pointer of a different project is passed to `processSegment`. Probably our flawless has been to create the implementation to actually have a plan of how to test it, so I see (now) that we could end up with an implementation easy to test if we would have started writing the test.

The test helper functions for generating values may have bugs (like a static test can have, but these ones have more chances), the only thing that makes here the way to detect a bug, in the test or the implementation, it's that each one has been implemented by a different person and if a test fails, we have to find out where the bug is. Apart from that, the random value generation makes harder than a bug in the test helper functions exist equally in the implementation because on each run the values and combination of them change wich drop the likelihood that we don't spot the bug (wherever it's).

We've already found bugs with this random test values generation that I didn't think of having a test case; the one that I've remember is the case for the 65 segments object.

Not saying that it isn't a good idea to have specific test cases, there are actually some of them with minor random value generation (just values, not combinations, which are the ones that are complex), but it was for me a way to kind of "black box" (it isn't really that because it's checking internal state) an implementation that existed knowing that `processSegment` will be called with all the `metainfoDB` segments and that's quite unpredictable to me.